controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -589,17 +595,26 @@ func (r *KubeadmControlPlaneReconciler) reconcileExternalReference(ctx context.C
 		return err
 	}
 
+	expectedOwnerRef := metav1.OwnerReference{
+		APIVersion: clusterv1.GroupVersion.String(),
+		Kind:       "Cluster",
+		Name:       cluster.Name,
+		UID:        cluster.UID,
+	}
+
+	if util.IsPaused(cluster, obj) && !util.HasOwnerRef(obj.GetOwnerReferences(), expectedOwnerRef) {
+		return errors.Wrapf(&capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second},
+			"Referenced %s %s is paused without an owner reference set.",
+			obj.GetKind(),
+			obj.GetName())
+	}
+
 	patchHelper, err := patch.NewHelper(obj, r.Client)
 	if err != nil {
 		return err
 	}
 
-	obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), metav1.OwnerReference{
-		APIVersion: clusterv1.GroupVersion.String(),
-		Kind:       "Cluster",
-		Name:       cluster.Name,
-		UID:        cluster.UID,
-	}))
+	obj.SetOwnerReferences(util.EnsureOwnerRef(obj.GetOwnerReferences(), expectedOwnerRef))
```

Comment:
Can we use controller-runtime utils to add this reference?

Reply:
Maybe? controller-runtime only contains a method for setting an owner reference that has the Controller flag set `SetControllerReference`.

This has me reviewing the fact we are setting a controller ref on the bootstrap and infrastructure resources we are creating alongside the Machines, and I'm thinking we should likely only be setting an ownerRef there (where the Machine Controller should likely have a controller ref rather than KCP).

With that said, I do think it's likely a good idea to turn this into a Controller reference here :)

Reply:
@vincepri thanks for catching that. I've fixed up the use of OwnerReferences within the KubeadmControlPlane controller and also patched up the external reference handling for Machine/Cluster controllers as well.